### ISSUE TITLE
Investigate memory issue of k_order dll (Dynare++)

### ISSUE THREAD

**Issue by JohannesPfeifer** [DynareTeam/dynare#1026]
See http://www.dynare.org/phpBB3/viewtopic.php?f=1&t=7155. Even with 32GB RAM, it crashes with 
```terminate called after throwing an instance of 'std::bad_alloc'
what(): std::bad_alloc```
which seems weird. This may also be related to #612

**Comment by MichelJuillard** [comment]
The model is rather large with 28 variables, 2 shocks and 11 state variables. I' m even surprised that it turns for order=6 on 32GB. Cross-derivatives of higher orders take a lot of space!

**Comment by JohannesPfeifer** [comment]
@MichelJuillard I am not really satisfied by your answer. Both the user and I checked with different OS and the resource monitor does not show anything suspicious. Under Windows 8, Dynare++ tells me right from the start in the ```jnl``` file
```dim=2 run out of memory```
It then continues until the crash at 
``` 
1627.46:M01675:-1.:-0.00:      :       dim=6 run out of memory, imposing max=10 (decrease number of threads)
1640.16:M01676:-1.:-0.00:      :       dim=6 avmem=0 tmpmem=756 max=10 stacks=4->6
1640.16:E01664:-1.:-0.00:000000:     Faa Di Bruno Z container for [0,0,4,2]
1640.83:E01596:-1.:-0.00:000000:   Recovering symmetry [0,0,0,6]
1640.83:E00905:-1.:-0.00:000000: Performing step for order = 6
```
with occasional large hard-disk activity, but hardly any RAM use. The used RAM increases during the run by maybe 200-500MB. But that leaves still about 27GB of free memory. So for now I don't see this as a problem with too large a model, particularly given our history with memory issues with the k_order_pert-dll (#612)

**Comment by MichelJuillard** [comment]
The dynamic model has 42 variables + 2 shocks and 28 equations. With the total number of variables n=44,
the number of non-symmetric derivatives for order 7 is
28*n*(n+1)*(n+2)*(n+3)*(n+4)*(n+5)*(n+6)/7 which is a very large number, even if dynare++ stores only nonzero derivatives. In addition, you need the information for orders < 7 and for storing the solution.
Most likely, dynare++ computes all the derivatives of the dynamic model before starting to compute the solution order by order (what you see in the journal)

**Comment by JohannesPfeifer** [comment]
I understand the intuitive argument, but it hardly seems to match what's going on on the machine. On a 64 bit system, there is no effective limitation on memory use as the harddisk should be used as swap space. But Dynare++ does not even use the full physical memory at any point in time. This in addition with the always wrong display of memory used makes me skeptical that just size is the problem.

**Comment by MichelJuillard** [comment]
If you want to investigate the issue further, you should probably do it on a Linux machine that is likely to give you a more accurate picture of instantaneous memory usage.. The basic fact remains that there are some problems that dynare++ can solve in a given among of RAM and some bigger problems that it can't. I believe that improving the memory efficiency of dynare++ would be a major work for little additional benefits as memory requirements increase very fast with the order of approximation..

**Comment by JohannesPfeifer** [comment]
I am not really interested in improving the memory efficiency. I worry more that there is a bigger issue here on non-Linux systems. Ticket #612, which was a memory leak under Windows, was closed without a solution. Now there is another case in the same basic routine where the memory management under Windows behaves counterintuitively.

**Comment by MichelJuillard** [comment]
You' are right, there maybe another problem. I tried the example on a Linux machine and dynare++ crashed after 25 minutes with a segmentation fault. According to the journal, it occurred when dealing with order=6. Until then the peak memory used was about 8GB

**Comment by JohannesPfeifer** [comment]
Do we then still view this issue as closed? And if not, how should we proceed?

**Comment by MichelJuillard** [comment]
gdb reported the error as occuring in openblas, but I doubt that openblas be th culprit. I'm trying valgrind but don' t expect a result before tomorrow

**Comment by JohannesPfeifer** [comment]
When building Dynare++ for Windows, do we distinguish 64 and 32 bit systems?

**Comment by houtanb** [comment]
No, we distribute a 32bit binary

**Comment by JohannesPfeifer** [comment]
Is Linux different? That might explain the problems in memory management under Windows where the out of memory happens straight from the start. For 32 bit the limit should be somewhere around 3GB memory. 
The only strange thing is that something similar happens for the ```k_order_pert.mex``` which is compiled for 64 bit.

**Comment by houtanb** [comment]
I imagine @MichelJuillard compiled ```dynare++``` in which case it would be the 64-bit version

**Comment by JohannesPfeifer** [comment]
What happens on Mac OS? The user reported the same message as under Windows, while Michel reports something different.

**Comment by houtanb** [comment]
On OS X and Windows we distribute a 32-bit version of dynare++. If the user compiled it himself, it could be either 32-bit or 64-bit, depending on his setup

**Comment by stepan-a** [comment]
Won't fix.

**Comment by JohannesPfeifer** [comment]
@stepan-a and @MichelJuillard Sorry, but I think we have to reopen the issue. I encounter the same problem on Windows at ```order=4``` on a 32GB machine, while I can run the same file flawlessly with ```order=5``` on Sedna.

**Comment by MichelJuillard** [comment]
We should reopen the issue, but currently we don't have the environment to investigate it. We would need to install a full Dynare development (compilation and debugging) environment on a Windows machine. I don't know who could do it and where.
